File: src/main/java/org/jcodec/codecs/vpx/vp9/Residual.java
package org.jcodec.codecs.vpx.vp9;

import static org.jcodec.codecs.vpx.vp9.Consts.ADST_DCT;
import static org.jcodec.codecs.vpx.vp9.Consts.BLOCK_16X16;
import static org.jcodec.codecs.vpx.vp9.Consts.BLOCK_16X32;
import static org.jcodec.codecs.vpx.vp9.Consts.BLOCK_16X8;
import static org.jcodec.codecs.vpx.vp9.Consts.BLOCK_32X16;
import static org.jcodec.codecs.vpx.vp9.Consts.BLOCK_32X32;
import static org.jcodec.codecs.vpx.vp9.Consts.BLOCK_32X64;
import static org.jcodec.codecs.vpx.vp9.Consts.BLOCK_4X4;
import static org.jcodec.codecs.vpx.vp9.Consts.BLOCK_4X8;
import static org.jcodec.codecs.vpx.vp9.Consts.BLOCK_64X32;
import static org.jcodec.codecs.vpx.vp9.Consts.BLOCK_64X64;
import static org.jcodec.codecs.vpx.vp9.Consts.BLOCK_8X16;
import static org.jcodec.codecs.vpx.vp9.Consts.BLOCK_8X4;
import static org.jcodec.codecs.vpx.vp9.Consts.BLOCK_8X8;
import static org.jcodec.codecs.vpx.vp9.Consts.BLOCK_INVALID;
import static org.jcodec.codecs.vpx.vp9.Consts.DCT_ADST;
import static org.jcodec.codecs.vpx.vp9.Consts.DCT_VAL_CAT6;
import static org.jcodec.codecs.vpx.vp9.Consts.PARETO_TABLE;
import static org.jcodec.codecs.vpx.vp9.Consts.TOKEN_TREE;
import static org.jcodec.codecs.vpx.vp9.Consts.TX_4X4;
import static org.jcodec.codecs.vpx.vp9.Consts.ZERO_TOKEN;
import static org.jcodec.codecs.vpx.vp9.Consts.blH;
import static org.jcodec.codecs.vpx.vp9.Consts.blW;
import static org.jcodec.codecs.vpx.vp9.Consts.cat_probs;
import static org.jcodec.codecs.vpx.vp9.Consts.coefband_4x4;
import static org.jcodec.codecs.vpx.vp9.Consts.coefband_8x8plus;
import static org.jcodec.codecs.vpx.vp9.Consts.extra_bits;
import static org.jcodec.codecs.vpx.vp9.Consts.maxTxLookup;

import org.jcodec.codecs.vpx.VPXBooleanDecoder;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * @author The JCodec project
 * 
 */
public class Residual {

    private int[][][] coefs;

    public Residual(int[][][] coefs) {
        this.coefs = coefs;
    }
    
    public static Residual read(int miCol, int miRow, int blSz, VPXBooleanDecoder decoder, Probabilities probStore,
            DecodingContext c, ModeInfo mode) {
        int[][][] coefs = new int[3][][];
        for (int pl = 0; pl < 3; pl++) {
            int subW = msb(blW[blSz] / c.getSubX()) - 2;
            int subH = msb(blH[blSz] / c.getSubY()) - 2;
            int uvBlSz = blk_size_lookup[subH][subW - subH + 1];
            int txSize = pl == 0 ? mode.getTxSize() : uvTxSize(c, mode.getTxSize(), blSz, uvBlSz);
            int step4x4 = 1 << txSize;
            int plBlSz = pl == 0 ? blSz : uvBlSz;

            int frameWPix = (c.getMiFrameWidth() << 3) >> c.getSubX();
            int frameHPix = (c.getMiFrameHeight() << 3) >> c.getSubY();
            int blX = (miCol << 3) >> c.getSubX();
            int blY = (miRow << 3) >> c.getSubY();

            coefs[pl] = new int[blH[plBlSz] * blW[plBlSz]][];
            for (int y = 0, blkIdx = 0; y < blH[plBlSz]; y += step4x4) {
                for (int x = 0; x < blW[plBlSz]; x += step4x4, blkIdx++) {
                    int posX = blX + (x << 2);
                    int posY = blY + (y << 2);
                    if (!mode.isSkip() && posX < frameWPix && posY < frameHPix) {
                        coefs[pl][blkIdx] = tokens(pl, posX, posY, txSize, blkIdx, mode.isInter(), decoder, probStore,
                                c);
                    }
                }
            }
        }
        return new Residual(coefs);
    }

    private static int msb(int v) {
        v &= 0xff;
        if ((v & 0xf0) != 0) {
            if ((v & 0xc0) != 0) {
                return 6 | (v >> 7);
            } else {
                return 4 | (v >> 5);
            }
        } else {
            if ((v & 0xc) != 0) {
                return 2 | (v >> 3);
            } else {
                return v >> 1;
            }
        }
    }

    public static int[][] blk_size_lookup = new int[][] { { BLOCK_INVALID, BLOCK_4X4, BLOCK_8X4 },
            { BLOCK_4X8, BLOCK_8X8, BLOCK_16X8 }, { BLOCK_8X16, BLOCK_16X16, BLOCK_32X16 },
            { BLOCK_16X32, BLOCK_32X32, BLOCK_64X32 }, { BLOCK_32X64, BLOCK_64X64, BLOCK_INVALID }, };

    private static int uvTxSize(DecodingContext c, int txSize, int blSz, int uvBlSz) {
        if (blSz < BLOCK_8X8)
            return TX_4X4;

        return Math.min(txSize, maxTxLookup[uvBlSz]);
    }

    public static int[] tokens(int plane, int startX, int startY, int txSz, int blockIdx, boolean isInter,
            VPXBooleanDecoder decoder, Probabilities probStore, DecodingContext c) {
        int maxCoeff = 16 << (txSz << 1);
        boolean expectMoreCoefs = true;
        int[] scan = c.getScan(plane, txSz, blockIdx);
        int txType = c.getTxType(plane, txSz, blockIdx);
        int[] coefs = new int[maxCoeff];
        for (int cf = 0; cf < maxCoeff; cf++) {
            int band = (txSz == TX_4X4) ? coefband_4x4[cf] : coefband_8x8plus[cf];
            int pos = scan[cf];
            if (!expectMoreCoefs) {
                boolean moreCoefs = readMoreCoefs(plane, pos, txSz, startX, startY, txType, band, isInter, decoder,
                        probStore, c);
                if (!moreCoefs)
                    break;
            }
            int token = readToken(plane, pos, txSz, startX, startY, txType, band, isInter, decoder, probStore, c);
            if (token == ZERO_TOKEN) {
                expectMoreCoefs = true;
                coefs[pos] = 0;
            } else {
                int coef = readCoef(token, decoder, c);
                int sign = decoder.readBitEq();
                coefs[pos] = sign == 1 ? -coef : coef;
                expectMoreCoefs = false;
            }
        }
        return coefs;
    }

    private static int readCoef(int token, VPXBooleanDecoder decoder, DecodingContext c) {
        int cat = extra_bits[token][0];
        int numExtra = extra_bits[token][1];
        int coef = extra_bits[token][2];
        if (token == DCT_VAL_CAT6) {
            for (int bit = 0; bit < c.getBitDepth() - 8; bit++) {
                int high_bit = decoder.readBit(255);

                coef += high_bit << (5 + c.getBitDepth() - bit);
            }
        }
        for (int bit = 0; bit < numExtra; bit++) {
            int coef_bit = decoder.readBit(cat_probs[cat][bit]);

            coef += coef_bit << (numExtra - 1 - bit);
        }
        return coef;
    }

    private static int pareto(int bin, int prob) {
        if (bin < 2) {
            return prob;
        }
        int x = (prob - 1) / 2;
        if ((prob & 1) != 0)
            return PARETO_TABLE[x][bin - 2];
        else
            return (PARETO_TABLE[x][bin - 2] + PARETO_TABLE[x + 1][bin - 2]) >> 1;
    }

    private static int readToken(int plane, int coefi, int txSz, int posX, int posY, int txType, int band,
            boolean isInter, VPXBooleanDecoder decoder, Probabilities probStore, DecodingContext c) {
        int ctx = calcTokenContext(plane, coefi, txSz, posX, posY, txType, c);
        int[][][][][][] probs = probStore.getCoefProbs();
        int prob0 = pareto(0, probs[txSz][plane > 0 ? 1 : 0][isInter ? 1 : 0][band][ctx][1]);
        int prob1 = pareto(1, probs[txSz][plane > 0 ? 1 : 0][isInter ? 1 : 0][band][ctx][2]);
        return decoder.readTree(TOKEN_TREE, prob0, prob1);
    }

    private static boolean readMoreCoefs(int plane, int coefi, int txSz, int posX, int posY, int txType, int band,
            boolean isInter, VPXBooleanDecoder decoder, Probabilities probStore, DecodingContext c) {
        int ctx = calcTokenContext(plane, coefi, txSz, posX, posY, txType, c);
        int[][][][][][] probs = probStore.getCoefProbs();

        return decoder.readBit(probs[txSz][plane > 0 ? 1 : 0][isInter ? 1 : 0][band][ctx][0]) == 1;
    }

    private static int calcTokenContext(int plane, int coefi, int txSz, int posX, int posY, int txType,
            DecodingContext c) {
        if (coefi == 0) {
            int[][] aboveNonzeroContext = c.getAboveNonzeroContext();
            int[][] leftNonzeroContext = c.getLeftNonzeroContext();
            int subX = plane > 0 ? c.getSubX() : 0;
            int subY = plane > 0 ? c.getSubY() : 0;
            int max4x = (c.getMiFrameWidth() << 1) >> subX;
            int max4y = (c.getMiFrameHeight() << 1) >> subY;
            int tx4 = 1 << txSz;
            int pos4x = posX >> 2;
            int pos4y = posY >> 2;
            int aboveNz = 0;
            int leftNz = 0;
            for (int i = 0; i < tx4; i++) {
                if (pos4x + i < max4x)
                    aboveNz |= aboveNonzeroContext[plane][pos4x + i];
                if (pos4y + i < max4y)
                    leftNz |= leftNonzeroContext[plane][pos4y + i];
            }
            return aboveNz + leftNz;
        } else {
            int abovePos = 0;
            int leftPos = 0;
            if (coefi != 0) {
                txSz += 2;
                int y = coefi >> txSz;
                int x = coefi & (0x3f >> (6 - txSz));
                abovePos = ((y - 1) << txSz) + x;
                leftPos = (y << txSz) + x - 1;
                if (txType == DCT_ADST || x == 0) {
                    leftPos = abovePos;
                }
                if (txType == ADST_DCT || y == 0) {
                    abovePos = leftPos;
                }
            }

            int[] tokenCache = c.getTokenCache();

            return (1 + tokenCache[abovePos] + tokenCache[leftPos]) >> 1;
        }
    }

}

File: src/main/java/org/jcodec/codecs/vpx/vp9/Residual.kt
package org.jcodec.codecs.vpx.vp9

import org.jcodec.codecs.vpx.VPXBooleanDecoder
import org.jcodec.codecs.vpx.vp9.Consts.BLOCK_16X16
import org.jcodec.codecs.vpx.vp9.Consts.BLOCK_16X32
import org.jcodec.codecs.vpx.vp9.Consts.BLOCK_16X8
import org.jcodec.codecs.vpx.vp9.Consts.BLOCK_32X16
import org.jcodec.codecs.vpx.vp9.Consts.BLOCK_32X32
import org.jcodec.codecs.vpx.vp9.Consts.BLOCK_32X64
import org.jcodec.codecs.vpx.vp9.Consts.BLOCK_4X4
import org.jcodec.codecs.vpx.vp9.Consts.BLOCK_4X8
import org.jcodec.codecs.vpx.vp9.Consts.BLOCK_64X32
import org.jcodec.codecs.vpx.vp9.Consts.BLOCK_64X64
import org.jcodec.codecs.vpx.vp9.Consts.BLOCK_8X16
import org.jcodec.codecs.vpx.vp9.Consts.BLOCK_8X4
import org.jcodec.codecs.vpx.vp9.Consts.BLOCK_8X8
import org.jcodec.codecs.vpx.vp9.Consts.BLOCK_INVALID
import org.jcodec.codecs.vpx.vp9.Consts.DCT_VAL_CAT1
import org.jcodec.codecs.vpx.vp9.Consts.DCT_VAL_CAT3
import org.jcodec.codecs.vpx.vp9.Consts.PARETO_TABLE
import org.jcodec.codecs.vpx.vp9.Consts.TOKEN_TREE
import org.jcodec.codecs.vpx.vp9.Consts.TX_4X4
import org.jcodec.codecs.vpx.vp9.Consts.blH
import org.jcodec.codecs.vpx.vp9.Consts.blW
import org.jcodec.codecs.vpx.vp9.Consts.cat_probs
import org.jcodec.codecs.vpx.vp9.Consts.coefband_4x4
import org.jcodec.codecs.vpx.vp9.Consts.coefband_8x8plus
import org.jcodec.codecs.vpx.vp9.Consts.extra_bits

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * @author The JCodec project
 */
open class Residual {
    var coefs: Array<Array<IntArray?>?> = emptyArray()
        private set

    constructor(coefs: Array<Array<IntArray?>?>) {
        this.coefs = coefs
    }

    constructor() {}

    fun read(miCol: Int, miRow: Int, blType: Int, decoder: VPXBooleanDecoder?, c: DecodingContext,
             modeInfo: ModeInfo) {
        if (modeInfo.isSkip) return
        val subXRound = (1 shl c.subX) - 1
        val subYRound = (1 shl c.subY) - 1
        val coefs: Array<Array<IntArray?>?> = arrayOfNulls(3)
        for (pl in 0..2) {
            val txSize = if (pl == 0) modeInfo.txSize else Consts.uv_txsize_lookup[blType][modeInfo.txSize][c.subX][c.subY]
            val step4x4 = 1 shl txSize
            var n4w = 1 shl blW.get(blType)
            var n4h = 1 shl blH.get(blType)
            if (pl != 0) {
                n4w = n4w shr c.subX
                n4h = n4h shr c.subY
            }
            var extra4w = (miCol shl 1) + n4w - (c.getFrameWidth() + 3 shr 2)
            var extra4h = (miRow shl 1) + n4h - (c.getFrameHeight() + 3 shr 2)
            var startBlkX = miCol shl 1
            var startBlkY = miRow shl 1
            if (pl != 0) {
                extra4w = extra4w + subXRound shr c.subX
                extra4h = extra4h + subYRound shr c.subY
                startBlkX = startBlkX shr c.subX
                startBlkY = startBlkY shr c.subY
            }
            val max4w = n4w - if (extra4w > 0) extra4w else 0
            val max4h = n4h - if (extra4h > 0) extra4h else 0
            coefs[pl] = arrayOfNulls(n4w * n4h)
            var y = 0
            while (y < max4h) {
                var x = 0
                while (x < max4w) {
                    val blkCol = startBlkX + x
                    val blkRow = startBlkY + y
                    var predMode: Int
                    if (pl == 0) {
                        predMode = modeInfo.yMode
                        if (blType < BLOCK_8X8) predMode = ModeInfo.vect4get(modeInfo.subModes, (y shl 1) + x)
                    } else predMode = modeInfo.uvMode
                    coefs[pl]!![x + n4w * y] = readOneTU(if (pl == 0) 0 else 1, blkCol, blkRow, txSize, modeInfo.isInter,
                            predMode, decoder, c)
                    x += step4x4
                }
                y += step4x4
            }
        }
        this.coefs = coefs
    }

    open fun readOneTU(plane: Int, blkCol: Int, blkRow: Int, txSz: Int, isInter: Boolean, intraMode: Int,
                       decoder: VPXBooleanDecoder?, c: DecodingContext): IntArray? {
        val decoder = decoder!!
        val tokenCache = IntArray(16 shl (txSz shl 1))
        val maxCoeff = 16 shl (txSz shl 1)
        var expectMoreCoefs = false
        val txType = if (plane == 0 && !isInter) Consts.intra_mode_to_tx_type_lookup[intraMode] else Consts.DCT_DCT
        val scan = if (plane == 0 && !isInter) Scan.vp9_scan_orders[txSz][txType][0] else Scan.vp9_default_scan_orders[txSz][0]
        val neighbors = if (plane == 0 && !isInter) Scan.vp9_scan_orders[txSz][txType][2] else Scan.vp9_default_scan_orders[txSz][2]
        val coefs = IntArray(maxCoeff)
        var ctx = calcTokenContextCoef0(plane, txSz, blkCol, blkRow, c)
        for (cf in 0 until maxCoeff) {
            val band: Int = if (txSz == TX_4X4) coefband_4x4.get(cf) else coefband_8x8plus.get(cf)
            val pos = scan[cf]
            val probs = c.getCoefProbs()[txSz][if (plane > 0) 1 else 0][if (isInter) 1 else 0][band][ctx]
            if (!expectMoreCoefs) {
                val moreCoefs = decoder.readBit(probs[0]) == 1
                if (!moreCoefs) break
            }
            var coef: Int
            if (decoder.readBit(probs[1]) == 0) {
                tokenCache[pos] = 0
                expectMoreCoefs = true
            } else {
                expectMoreCoefs = false
                if (decoder.readBit(probs[2]) == 0) {
                    tokenCache[pos] = 1
                    coef = 1
                } else {
                    val token = decoder.readTree(TOKEN_TREE, PARETO_TABLE.get(probs[2] - 1))
                    if (token < DCT_VAL_CAT1) {
                        coef = token
                        if (token == Consts.TWO_TOKEN) tokenCache[pos] = 2 else tokenCache[pos] = 3
                    } else {
                        if (token < DCT_VAL_CAT3) tokenCache[pos] = 4 else tokenCache[pos] = 5
                        coef = readCoef(token, decoder, c)
                    }
                }
                val sign = decoder.readBitEq()
                coefs[pos] = if (sign == 1) -coef else coef
            }
            ctx = 1 + tokenCache[neighbors[2 * cf + 2]] + tokenCache[neighbors[2 * cf + 3]] shr 1
            println("CTX: $ctx")
        }
        return coefs
    }

    companion object {
        @JvmStatic
        fun readResidual(miCol: Int, miRow: Int, blSz: Int, decoder: VPXBooleanDecoder, c: DecodingContext,
                         mode: ModeInfo): Residual {
            val ret = Residual()
            ret.read(miCol, miRow, blSz, decoder, c, mode)
            return ret
        }

        var blk_size_lookup = arrayOf(intArrayOf(BLOCK_INVALID, BLOCK_4X4, BLOCK_8X4), intArrayOf(BLOCK_4X8, BLOCK_8X8, BLOCK_16X8), intArrayOf(BLOCK_8X16, BLOCK_16X16, BLOCK_32X16), intArrayOf(BLOCK_16X32, BLOCK_32X32, BLOCK_64X32), intArrayOf(BLOCK_32X64, BLOCK_64X64, BLOCK_INVALID))
        private fun readCoef(token: Int, decoder: VPXBooleanDecoder, c: DecodingContext): Int {
            val cat: Int = extra_bits.get(token).get(0)
            val numExtra: Int = extra_bits.get(token).get(1)
            var coef: Int = extra_bits.get(token).get(2)
            // if (token == DCT_VAL_CAT6) {
            // for (int bit = 0; bit < c.getBitDepth() - 8; bit++) {
            // int high_bit = decoder.readBit(255);
            //
            // coef += high_bit << (5 + c.getBitDepth() - bit);
            // }
            // }
            for (bit in 0 until numExtra) {
                val coef_bit = decoder.readBit(cat_probs.get(cat).get(bit))
                coef += coef_bit shl numExtra - 1 - bit
            }
            return coef
        }

        private fun calcTokenContextCoef0(plane: Int, txSz: Int, blkCol: Int, blkRow: Int, c: DecodingContext): Int {
            val aboveNonzeroContext = c.getAboveNonzeroContext()
            val leftNonzeroContext = c.getLeftNonzeroContext()
            val subX = if (plane > 0) c.subX else 0
            val subY = if (plane > 0) c.subY else 0
            val max4x = c.miFrameWidth shl 1 shr subX
            val max4y = c.miFrameHeight shl 1 shr subY
            val tx4 = 1 shl txSz
            var aboveNz = 0
            var leftNz = 0
            for (i in 0 until tx4) {
                if (blkCol + i < max4x) aboveNz = aboveNz or aboveNonzeroContext[plane][blkCol + i]
                if (blkRow + i < max4y) leftNz = leftNz or leftNonzeroContext[plane][blkRow + i and 0xf]
            }
            return aboveNz + leftNz
        }
    }
}